### PR TITLE
React to dnx confguration change (prototype)

### PR DESCRIPTION
**Pull request by pakrym**
@davidfowl @pranavkm

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs [hunk 1]
```diff
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using Microsoft.AspNet.FileProviders;
+using Microsoft.AspNet.Hosting;
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Shouldn't be needed.

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs [hunk 2]
```diff
@@ -42,5 +45,26 @@ public IFileProvider FileProvider
...
+        /// At startup, this is initialized to "Debug" if <see cref="Microsoft.AspNet.Hosting.IHostingEnvironment"/> service is
+        /// registred and environment is development (<see cref="HostingEnvironmentExtensions.IsDevelopment"/>) else it is set to
+        /// "Release".
+        /// </remarks>
+        public string Configuration
+        {
+            get { return _configuration; }
+            set
+            {
+                if (value == null)
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: Make this `string.IsNullOrEmpty` Generally use that to null checks for string values.

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs [hunk 3]
```diff
@@ -17,16 +18,26 @@ public class RazorViewEngineOptionsSetup : ConfigureOptions<RazorViewEngineOptio
...
         }
 
         private static void ConfigureRazor(
             RazorViewEngineOptions razorOptions,
-            IApplicationEnvironment applicationEnvironment)
+            IApplicationEnvironment applicationEnvironment,
+            IHostingEnvironment hostingEnvironment)
         {
             razorOptions.FileProvider = new PhysicalFileProvider(applicationEnvironment.ApplicationBasePath);
+            if (hostingEnvironment != null && hostingEnvironment.IsDevelopment())
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: Can IHostingEnvironment really be null?

Comment (pakrym) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: Yes, if someone want to use razor in console app

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs [hunk 4]
```diff
@@ -17,16 +18,26 @@ public class RazorViewEngineOptionsSetup : ConfigureOptions<RazorViewEngineOptio
...
+            {
+                razorOptions.Configuration = "Release";
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: Looks like the values are inverted for the case (`IsDevelopment` -> `Debug` else `Release`).

File: test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/RoslynCompilationServiceTest.cs [hunk 5]
```diff
@@ -19,6 +19,8 @@ namespace Microsoft.AspNet.Mvc.Razor.Compilation
 {
     public class RoslynCompilationServiceTest
     {
+        private const string _configurationName = "Release";
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Razor.Test/Compilation/RoslynCompilationServiceTest.cs]: `private const string ConfigurationName`

File: src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs [hunk 6]
```diff
@@ -38,9 +42,14 @@ public static class CompilationOptionsProviderExtension
                 throw new ArgumentNullException(nameof(applicationEnvironment));
             }
 
+            if (configuration == null)
+            {
+                throw new ArgumentNullException(nameof(configuration));
+            }
+
             return compilerOptionsProvider.GetCompilerOptions(applicationEnvironment.ApplicationName,
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/Compilation/CompilationOptionsProviderExtension.cs]: Completely unrelated - while you're modifying this type, could you reformat this bit of code so it uses the VS default indentation:
```C#
return compilerOptionsProvider.GetCompilerOptions(
        applicationEnvironment.ApplicationName,
        applicationEnvironment.RuntimeFramework,
        configuration)
    .ToCompilationSettings(applicationEnvironment.RuntimeFramework, applicationEnvironment.ApplicationBasePath);
```

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs (hunk 3 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: That makes no sense. This is the RazorViewEngine, not the Razor parser.

Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: We assume everything passed in to the `ctor` is non null. You could remove the null check.

Comment (pakrym) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptionsSetup.cs]: Will do

File: src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs [hunk 7]
```diff
@@ -42,5 +44,26 @@ public IFileProvider FileProvider
...
+        /// "Release".
+        /// </remarks>
+        public string Configuration
+        {
+            get { return _configuration; }
+            set
+            {
+                if (string.IsNullOrEmpty(value))
+                {
+                    throw new ArgumentNullException(nameof(value));
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Razor/RazorViewEngineOptions.cs]: `throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(value));`

**Comment by pranavkm** [comment]
:shipit:

**Comment by pakrym** [comment]
I added more tests, please have a look

File: test/Microsoft.AspNet.Mvc.Test/RazorViewEngineOptionsSetupTest.cs [hunk 8]
```diff
@@ -29,5 +33,45 @@ public void RazorViewEngineOptionsSetup_SetsUpFileProvider()
             Assert.NotNull(options.FileProvider);
             Assert.IsType<PhysicalFileProvider>(options.FileProvider);
         }
+
+        [Fact]
+        public void RazorViewEngineOptionsSetup_SetsConfigurationDebugForDevelopmentEnv()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Test/RazorViewEngineOptionsSetupTest.cs]: `DevelopmentEnvironment` (no need to abbreviate it)

File: test/Microsoft.AspNet.Mvc.Test/RazorViewEngineOptionsSetupTest.cs [hunk 9]
```diff
@@ -29,5 +33,45 @@ public void RazorViewEngineOptionsSetup_SetsUpFileProvider()
...
+
+            // Act
+            optionsSetup.Configure(options);
+
+            // Assert
+            Assert.Equal("Debug", options.Configuration);
+        }
+
+        [Fact]
+        public void RazorViewEngineOptionsSetup_SetsConfigurationReleaseForNonDevelopmentEnv()
```
Comment (pranavkm) [inline test/Microsoft.AspNet.Mvc.Test/RazorViewEngineOptionsSetupTest.cs]: Could you make this a TheoryData and include `Production` and `Staging` here?

Comment (pakrym) [inline test/Microsoft.AspNet.Mvc.Test/RazorViewEngineOptionsSetupTest.cs]: Maybe one test that takes environment and expected configuration?

**Pull request merged**